ruby_oop/mastermind/game.rb

```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+require './display_text'
+
+# class to play the mastermind game
+class Game
+  include DisplayableText
+
+  AVAILABLE_COLORS = ['🟤', '🔴', '🔵', '🟢', '🟡', '🟣'].freeze
+
+  def initialize
+    @player_codebreaker = codebreaker_or_codemaker?
+  end
+
+  # keep playing until there's a match OR 12 turns happen
+  def play
+    turn_number = 0
+    if @player_codebreaker
+      human_codemaker
+      until game_over? || turn_number == 12
+        computer_codebreaker
+        color_index_match
+        color_found_in_array
+        turn_number += 1
+      end
+    else
+      computer_codemaker
+      until game_over? || turn_number == 12
+        human_codebreaker
+        color_index_match
+        color_found_in_array
+        turn_number += 1
+        puts 'Try again.'
+      end
+    end
+  end
+
+  private
+
+  def game_over?
+    @game_over = false
+    if win?
+      puts "Game over! Exact match. The code was: #{@secret_code}."
+      @game_over = true
+    else
+      @game_over = false
+    end
+    @game_over
+  end
+
+  def win?
+    @guess_array == @secret_code
+  end
+
+  def pick_color(color)
+    {
+      'br' => '🟤',
+      'r' => '🔴',
+      'b' => '🔵',
+      'g' => '🟢',
+      'y' => '🟡',
+      'p' => '🟣'
+    }[color]
+  end
+
+  def color_choice_selection
+    puts display_color_choice_prompt
+    gets.chomp
+  end
+
+  def computer_codemaker
+    puts display_computer_text
+    @secret_code = AVAILABLE_COLORS.sample(4)
+  end
+
+  def computer_codebreaker
+    puts display_computer_text
+    @guess_array = AVAILABLE_COLORS.sample(4)
+  end
+
+  def human_codemaker
+    puts display_codemaker_confirmation_text
+    @secret_code = set_code_array
+  end
+
+  def human_codebreaker
+    puts display_codebreaker_confirmation_text
+    @guess_array = set_code_array
+    if @guess_array.include?(nil) || @guess_array.uniq.length != 4
+      puts display_human_codebreaker_display_dupe_text
+      @guess_array = set_code_array
+    end
+    puts "\nYou guessed: #{@guess_array}"
+    @guess_array
+  end
+
+  def set_code_array
+    [pick_color(color_choice_selection), pick_color(color_choice_selection),
+     pick_color(color_choice_selection), pick_color(color_choice_selection)]
+  end
+
+  def color_found_in_array
```

Comment:
I don't think it's worth the refactor at this point, but for future design considerations I would pull the `puts` out of methods like this and have it only be concerned with returning `true` or `false`, and renamed to `color_found_in_array?`. So it'd be called like `puts "blah blah" if color_found_in_array?`. Separates the logic concerns from the UI concerns (`puts` in this case). 

Otherwise we risk spaghetti code situations as the code is iterated on over time.